### PR TITLE
Unpin machine applications on destroy

### PR DESCRIPTION
**Pull request by SimonRichardson**
##  Description of change

During a destruction of a machine application, we should attempt to
remove any last references to ties for leadership pinning.

The normal destroy flow will attempt to remove the leadership pinnings
but will fail if there exists any pinnings. Using force will allow this to
proceed.

The call for removing pinnings is idempotent, so calling it multiple
times is fine.

## QA steps

Assumes bootstrapping has been performed.

### Regression

```sh
juju add-machine -n 1
juju remove-machine 0

juju deploy cs:percona-cluster-290 -n 10

# The following should fail
juju machines --format=json | jq ".machines | keys[]" | xargs -I% juju remove-machine % 

# The following should succeed
juju machines --format=json | jq ".machines | keys[]" | xargs -I% juju remove-machine % --force

# Check for errors.
juju debug-log
```

### Upgrade series Q&A

```sh
juju deploy cs:percona-cluster-290

juju upgrade-series prepare 12 eoan
juju remove-machine 12 # fails (shouldn't call the new code)
juju remove-machine 12 --force
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1879663

### CODE REVIEW

File: featuretests/remoterelations_test.go [hunk 1]
```diff
@@ -160,7 +160,7 @@ func (s *remoteRelationsSuite) TestWatchLocalRelationChanges(c *gc.C) {
 	assertRelationUnitsChange(c, s.BackingState, w, params.RemoteRelationChangeEvent{
 		RelationToken:    relToken,
 		ApplicationToken: appToken,
-		UnitCount: &uc,
+		UnitCount:        &uc,
```
Comment (SimonRichardson) [inline featuretests/remoterelations_test.go:163]: Linter is moaning about this for some reason!

File: mongo/mongo.go [hunk 2]
```diff
@@ -238,7 +238,7 @@ var (
 		Patch:         "",
 		StorageEngine: WiredTiger,
 	}
-	// MongoUpgrade represents a sepacial case where an upgrade is in
+	// MongoUpgrade represents a special case where an upgrade is in
```
Comment (SimonRichardson) [inline mongo/mongo.go:241]: Driveby: freaked me out.

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

File: apiserver/facades/client/machinemanager/machinemanager.go [hunk 3]
```diff
@@ -393,6 +436,19 @@ func (mm *MachineManagerAPI) destroyMachine(args params.Entities, force, keep bo
 				return fail(err)
 			}
 		}
+
+		// Ensure that when the machine has been removed that all the leadership
+		// references to that machine are also cleared up.
+		results, err := mm.leadership.UnpinApplicationLeadersByName(machineTag, applicationNames)
```
Comment (manadart) [inline apiserver/facades/client/machinemanager/machinemanager.go:442]: Not sure we can return a failure here.
The machine will have been removed, and we will tell the caller that we failed to remove it.

Comment (SimonRichardson) [reply apiserver/facades/client/machinemanager/machinemanager.go:442]: @manadart that's what I was wondering, didn't get to ask yesterday because of the commotion.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$
